mm2src/coins/z_coin/z_rpc.rs

```diff
@@ -43,6 +55,379 @@ struct CompactBlockRow {
+            for (tx_id, hash_tx) in block.tx.iter().enumerate() {
+                let tx_bytes = self.get_transaction_bytes(hash_tx).compat().await?;
+                let tx = ZTransaction::read(tx_bytes.as_slice()).unwrap();
+                let mut spends = Vec::new();
+                let mut outputs = Vec::new();
+                if !tx.shielded_spends.is_empty() || !tx.shielded_outputs.is_empty() {
+                    // Create and push spends with outs for compact_tx during iterations.
+                    for spend in &tx.shielded_spends {
+                        let compact_spend = TonicCompactSpend {
+                            nf: spend.nullifier.to_vec(),
+                        };
+                        spends.push(compact_spend);
+                    }
+                    for out in &tx.shielded_outputs {
+                        let compact_out = TonicCompactOutput {
+                            cmu: out.cmu.to_bytes().to_vec(),
+                            epk: out.ephemeral_key.to_bytes().to_vec(),
+                            // https://zips.z.cash/zip-0307#output-compression
+                            // The first 52 bytes of the ciphertext contain the contents and opening of the note commitment,
+                            // which is all of the data needed to spend the note and to verify that the note is spendable.
+                            ciphertext: out.enc_ciphertext[0..52].to_vec(),
+                        };
+                        outputs.push(compact_out);
+                    }
+                    // Shadowing mut variables as immutable. No longer need to update them.
+                    drop_mutability!(spends);
+                    drop_mutability!(outputs);
+                    let mut hash_tx_vec = hash_tx.0.to_vec();
+                    hash_tx_vec.reverse();
+
+                    let compact_tx = TonicCompactTx {
+                        index: tx_id as u64,
+                        hash: hash_tx_vec,
+                        fee: 0,
+                        spends,
+                        outputs,
+                    };
+                    compact_txs.push(compact_tx);
+                }
+            }
+            let mut hash = block.hash.0.to_vec();
+            hash.reverse();
+            // Set 0 in vector in the case of genesis block.
+            let mut prev_hash = block.previousblockhash.unwrap_or_default().0.to_vec();
+            prev_hash.reverse();
+            // Shadowing mut variables as immutable.
+            drop_mutability!(hash);
+            drop_mutability!(prev_hash);
+            drop_mutability!(compact_txs);
+
+            let compact_block = TonicCompactBlock {
+                proto_version: 0,
+                height,
+                hash,
+                prev_hash,
+                time: block.time,
+                // (hash, prevHash, and time) OR (full header)
+                header: Vec::new(),
```

Comment:
Is it safe to use default value here?

Reply:
Yeap, compiler understands, which type should be for proto format (u8 for bytes in header case).
I put an empty vector with a manually set type `i32` and got an error: expected `u8`, found `i32`. 
So compiler chooses u8 when we use `Vec::new()` for header.

Reply:
By "safe" I meant whether it can cause logical bugs, this have no relation to compiler behavior 🙂

Reply:
Right now light mode provides `CompactBlock`s with empty headers.
According to [documentation](https://zcash.readthedocs.io/en/latest/lightwalletd/index.html), we have to provide full header or hash, prevHash, and time.

In the future [Block header validation ](https://zips.z.cash/zip-0307#block-header-validation) could be implemented. Now only prevHash is checked for this enhancement. If this validation is implemented in light mode, we will have the logic bug. So we will need to implement the same validation.

Reply:
@laruh I see, is this field also empty when you get compact block data from lightwalletd?

Reply:
@artemii235 yeap, I checked compact blocks from lightwalletd in log, they always have `header: [],` field.